### ISSUE TITLE
New rule proposal: `prefer-appearance`

### ISSUE THREAD

**Issue by zaicevas** [testing-library/eslint-plugin-testing-library#373]
In a lot of cases, we're able to choose to:
- a) wait for appearance with async utility (waitFor/findBy)
- b) wait for disappearance with `waitForElementToBeRemoved` and then use synchronous query (e.g. getByText)

I am thinking we could have a rule to enforce one over the another 🤔  Example:

```
        // Incorrect
        await waitForElementToBeRemoved(() => screen.getByTestId('loader'))
        const postWithReply = screen.getByTestId('post-reply-item')

        // Correct
        const postWithReply = await screen.findByTestId('post-reply-item')
```

There are some more tricky cases:

```
      await waitForElementToBeRemoved(() => screen.getByTestId('loader'))
      userEvent.click(screen.getAllByText('Reply')[0]) // findAllByText won't work, AFAIC, but waitFor could be used
```

This is more of a idea-level proposal, I haven't dived deep into details :)

**Comment by Belco90** [comment]
Hey @zaicevas, thanks for your proposal!

It would be nice to report such a use case, but to be honest, it seems really tricky to report this correctly.

Some tricky scenarios:
- Knowing if `waitForElementToBeRemoved` and the following query are strictly related. It might be the case that `waitForElementToBeRemoved` is used to wait until some element disappears (like a modal) but the element being queried after is already there, so it wouldn't be possible to wait to find that element since is already there
  ```javascript
  // invalid? this seems legit to me
  await waitForElementToBeRemoved(() => screen.getByTestId('modal'))
  const somethingAlreadyThereBeforeOtherElementWasRemoved = screen.getByTestId('element-in-the-main-page')
  ```
- What if there are other calls between `waitForElementToBeRemoved` and the query? The example you added is too simple in that sense.
  ```javascript
  // invalid? I need to wait for all things to disappear before moving forward
  await waitForElementToBeRemoved(() => screen.getByTestId('loader'))
  await waitForElementToBeRemoved(() => screen.getByTestId('toast-message'))
  // other potential calls/actions here
  const postWithReply = screen.getByTestId('post-reply-item')
  ```
- What if there is no new element appearing in the UI? How do you know that the element being queried after the `waitForElementToBeRemoved` is a new element that appeared after an action? You can't assume that.

  ```javascript
  // invalid?
  const deleteButtonA = screen.getByTestId('delete-item-a')
  userEvent.click(deleteButtonA) // this deletes an item from a list but nothing new appears
  await waitForElementToBeRemoved(deleteButtonA)

  // now I want to query another existing button for deleting another item,
  // but this rule would complain about it,
  // forcing me to query this button BEFORE the waitForElementToBeRemoved
  // with no reason
  const deleteButtonB = screen.getByTestId('delete-item-b')
  userEvent.click(deleteButtonB)
  await waitForElementToBeRemoved(deleteButtonB)
  ```

I don't think this rule is doable in a real scenario since it has a heavyweight on user intentions.

**Comment by zaicevas** [comment]
Thanks for feedback! I think your 1st point (`the element being queried after is already there`) is enough to convince me that such rule is too tricky to implement, as we can't infer how the component is written :) Closing this issue.
